Add Quarkiverse blog entry (#773)
quarkusio/quarkusio.github.io: _posts/2020-11-17-quarkiverse.adoc

---
layout: post
title: Quarkiverse Hub - Home of Community Extensions
date: 2020-11-17
tags: extensions quarkiverse
synopsis: Entering a world of extensions
author: gastaldi
---

Quarkus is a Kubernetes Native Java stack tailored for the OpenJDK HotSpot and GraalVM, crafted from the best of breed Java libraries and standards, as well as an ever growing ecosystem of its extensions. At the beginning we were accepting all the contributed extensions in the core Quarkus repository. Eventually though it grew too large and lead to a maintenance overhead. In addition to that, in some cases it's simply not always sensible to include an extension into the core Quarkus repository.

Some have since then created their own repositories external to Quarkus, but others just never happened because not everyone wants the overhead of maintaining a build setup. Even the Quarkus core developers sometimes had ideas for an extension but not solid enough to be included in the Quarkus core repository.

Enter the *Quarkiverse*!

> image:/assets/images/posts/quarkiverse/quarkiverse.png[]

Reviewer: that image is too big :) 

Also, I don't see an explanation for what it is - i.e. what isn grok, what is the significance of the arrows etc. 

Thus, definitely make it smaller - and give some context.

Quarkiverse is everything that encompasses the Quarkus universe: that includes the core, platform and all extensions.

> When it comes to developing extensions maintained by the community, it is important to standardize the repository structure, the release process and provide an integration with the existing Quarkus tooling.

Reviewer: ```suggestion
When it comes to developing extensions it is simpler to collaborate if the repository structure, the release process and and integration with the existing Quarkus tooling is standardised and automated.
```

suggesting to remove "by the community" since it doesn't really matter  who does this

That's where Quarkiverse Hub comes into play.

== What is Quarkiverse Hub?

> The link:https://github.com/quarkiverse[Quarkiverse Hub] is our proposed solution to create a "home" for Quarkus extensions maintained by the community.

Reviewer: partial screenshot of quarkiverse hub front page with list of extensions would be nice here

By hosting an extension under the Quarkiverse GitHub organization, you'll get the following features:

- Automated and secured publishing of your Maven releases to Maven Central
- Automated cross-testing of your extension with Quarkus builds/releases (see https://github.com/quarkusio/quarkus-ecosystem-ci)
- Inclusion in the upcoming Extension registry used by Quarkus tooling to browse for extensions (i.e. https://code.quarkus.io, command line and IDE plugins).
- Consistent formatting and release of projects
- Documentation aggregation
- Bigger Community outreach and a great place to collaborate

> 

Reviewer: Maybe add something like: 

```suggestion

The advantage by this is that you as a extension developer do not need to register and setup Maven Central publishing nor sit around waiting for deployment to work - all you do is commit for next release and the automation does the rest.
```

?

== Requirements

> In order to ease the deployment and generate docs on the fly, there are some rules that need to be followed for projects under the Quarkiverse organization:

Reviewer: "generate docs on the fly" ...is that in place now ?

Author: Not generation, but publication. Will push a fix ASAP

> In order to ease the deployment and generate docs on the fly, there are some rules that need to be followed for projects under the Quarkiverse organization:

Reviewer: ```suggestion
In order to ease the deployment and automatically generate docs, there are some rules that need to be followed for projects under the Quarkiverse organization:
```

is that what you meant ?

Author: I think the right word is `publish` docs, not generate. I'll fix it

---
layout: post
title: Quarkiverse Hub - Home of Community Extensions
date: 2020-11-17
tags: extensions quarkiverse
synopsis: Entering a world of extensions
author: gastaldi
---

Today we're proud to announce link:https://github.com/quarkiverse[Quarkiverse Hub], a GitHub organization where any contributor can host their Quarkus extensions.

image:/assets/images/posts/quarkiverse/quarkiverse-repos.png[]

> Hosting your extension in the Quarkiverse Hub makes it available to all Quarkus users through the Quarkus tooling, and over time we'll start to add other features like a central extension registry.

Reviewer 2: I'm not sure people will know what is meant by a "central extension registry". I presume you are talking about something like npmjs.com that provides pages like https://www.npmjs.com/package/ssh2 per extension? I don't know that we can call out that comparison directly. However, we could maybe describe the aspects of it that would interest our users. Thinking about some text to propose...

---
layout: post
title: Quarkiverse Hub - Home of Community Extensions
date: 2020-11-17
tags: extensions quarkiverse
synopsis: Entering a world of extensions
author: gastaldi
---

> Quarkus is a Kubernetes Native Java stack tailored for the OpenJDK HotSpot and GraalVM, crafted from the best of breed Java libraries and standards, as well as an ever growing ecosystem of its extensions. At the beginning we were accepting all the contributed extensions in the core Quarkus repository. Eventually though it grew too large and lead to a maintenance overhead. In addition to that, in some cases it's simply not always sensible to include an extension into the core Quarkus repository.

Reviewer 2: I don't think it's a good idea to lead with this. It sounds to me like you are saying "there was an internal maintenance issue that the team needed to fix so we moved some stuff around. Here's a blog post on how to deal with the move"

I think it would be better to lead with some Tl;Dr; that excites the community. For example, something like: 

"Today we are announcing the Quarkiverse! A single repository where any contributor can host their Quarkus extension. Hosting your extension in the Quarkiverse makes it available to all Quarkus users through the Quarkus tooling, and over time we'll start to add other features like <insert some features we are willing to talk publicly about>. So read on to discover more and find out how you can make your extension available to all Quarkus developers today!"

Obviously that's a bit rough around the edges, and I'm likely not using all the correct terminology. But hopefully you get the idea.

Author: I agree this is much better, let me give it another try

Releasing an artifact is as simple as opening a pull-request changing the https://github.com/quarkiverse/quarkiverse-template/blob/master/.github/project.yml[`.github/project.yml`] file in the extension's repository.
Once the pull-request is merged, a GitHub action is triggered invoking `mvn release:prepare release:perform` and the extension is deployed to Maven central.

Automated cross-testing of your extension with Quarkus builds/releases::

Quarkus Core releases happen constantly and keeping up with the latest changes before a release is performed is crucial. Quarkiverse extensions are built daily against the latest https://github.com/quarkusio/quarkus[Quarkus core] using the
https://github.com/quarkusio/quarkus-ecosystem-ci[Quarkus Ecosystem CI] and listed in the https://status.quarkus.io/[Quarkus Status] page.

Consistent formatting and release of projects::

The `quarkiverse-parent` artifact (inherited by each extension in Quarkiverse) enables plugins that will format your source code to the same format used in Quarkus core.

Documentation aggregation::

> Quarkiverse Hub uses https://antora.org[Antora] to aggregate each extension documentation in the Quarkiverse docs website. Check it out at https://quarkiverse.github.io/quarkiverse-docs/

Reviewer: I would mark this as work in progress too ...not convinced antora actually will work...how is this handled when there is freemarker 1.0 and google-cloud 2.3 ?

Author: Antora already handles that, it's just a matter of configuring the `antora.yml` descriptor telling which versions to pick for each repository

For more information, look at the link:https://github.com/quarkiverse/quarkiverse/wiki[Quarkiverse Wiki]

image:/assets/images/posts/quarkiverse/quarkiverse-way.jpg[]

== How do I sign-up?

Create an *Extension Proposal* ticket in the link:https://github.com/quarkusio/quarkus/issues[Quarkus issue tracker] (first check if the extension you're proposing is not created yet) and let us know if you want to lead the development of this extension by commenting on the issue.

The Quarkus team will get you started with the boring stuff of setting up a brand new repository and deployment scripts, so you'll have fun coding it.

As usual, we're always improving, so take a look at link:https://github.com/quarkiverse/quarkiverse/wiki#joining-quarkiverse[Joining Quarkiverse] for the process we are currently using.

== What extensions are in Quarkiverse already?

Check out the link:https://github.com/search?q=topic%3Aquarkus-extension+org%3Aquarkiverse&type=Repositories[extensions that are already there].

> Stay tuned for the latest info about Quarkiverse in the link:https://github.com/quarkiverse/quarkiverse/wiki[Quarkiverse Wiki]

Reviewer: should we add anotice indicating that you can still have an project external to quarkus and quarkiverse - just that then there are more manual steps to do ?
